### ISSUE TITLE
Pushing changes before committing crashes application

### ISSUE THREAD

**Issue by omazhary** [uvicgse/project-virtual-team-1#30]
Using the push changes function after you’ve made a change to the repo but before you have committed crashes the application. This does not seem to happen every time but will happen eventually.

**Steps to reproduce:**
1. Edit a local file, but don't commit the change.

![image](https://user-images.githubusercontent.com/22312351/58006409-245f7280-7b3c-11e9-96a7-47d30d2c49a5.png)

2. Push changes (Without committed files)
![image](https://user-images.githubusercontent.com/22312351/58006744-deef7500-7b3c-11e9-9c8d-d075f856603b.png)

3. Application sometimes crashes (goes to black)
![image](https://user-images.githubusercontent.com/22312351/58006929-40afdf00-7b3d-11e9-831a-9b92b1ffd24d.png)

**Comment by colinwerner** [comment]
This is easy to reproduce, although it is slightly different as issue #47 needs to be fixed before we can see edited files (staged or unstaged).

My steps:
1.) Open local repository (doesn't matter if files are changed or not).
2.) Click push

Expected results:
3.) No Local commits: "You have not committed."
Local commits: "Files are committed" (assuming no merge conflicts).

Actual Results:
3.) Black Screen Of Death (BkSOD)

**Comment by jerushachua** [comment]
This is assigned to group 2.

**Comment by trstndvll** [comment]
![Screenshot 2019-05-22 16 44 34](https://user-images.githubusercontent.com/12837282/58216555-e1280f80-7cb3-11e9-99f2-13e3c1e8e3d7.jpg)

**Comment by trstndvll** [comment]
I added a small if statement in git.ts, where I added a check in pushToRemote() to check if we are attempting to push with no committed files. From what I can see on the console logs it's working, but I'm having a problem on Mac where I can only commit and push once. If I try and commit then push a second time, the app crashes. This is regardless if I take my changes in or out. Would another Mac user be able to confirm this behaviour? Or maybe a Windows user can test it as well? Not sure if my logic will work in every push situation but it seems to be on the right track.

**Comment by nikola-rados** [comment]
@trstndvll would you be able to share the code block that you are talking about? I'll be able to try it later on today.

**Comment by trstndvll** [comment]
@nikola-rados Yeah no problem! I did commit the changes to the branch, the block is in git.ts, lines 393-398, with a reference at line 105 as well
![Screenshot 2019-05-23 13 05 17](https://user-images.githubusercontent.com/12837282/58282725-6b777e80-7d5b-11e9-80c1-523e9f00a8e1.jpg)

**Comment by colinwerner** [comment]
I could actually reproduce the crashing issue, so here is what I did:

1.) opened a local repository
2.) edited two different files
3.) pushed the changes (not committed) -- I saw a message saying pushing, even though nothing was done.
4.) pushed the changes again and then the whole thing went kaput.

PS @trstndvll make sure to get a video of the original behaviour for demo, please!

**Comment by trstndvll** [comment]
Thanks @colinwerner I have a screen recording ready to go!

**Comment by colinwerner** [comment]
@trstndvll right on!

**Comment by trstndvll** [comment]
So I realized the check wasn't being reset at the end of the push function so I fixed that. Now it will alert you every time there's an empty push. Still having the issue with the app crashing every second (successful) commit and push though. Can the Windows/Linux folks speak to this?

**Comment by Unimax** [comment]
@trstndvll i notice if u commit changes and don't push. then restart the application. then it won't show any changes because they are committed but you cannot push because of your flag. I think we should update 'pushButNoCommit' using some git command to check committed files/staged files.

**Comment by trstndvll** [comment]
@Unimax thank you for pointing that out. I can confirm that behaviour. I've been trying out a few things over the past couple hours and I found that command "git cherry" will show you what local commits are not on the remote server. So, if the output of git cherry is nothing, that means we're shouldn't allow pushing. Otherwise, if git cherry returns a commit ID, then we know we should allow pushing. I'm just trying to find a way to take in the command line output as a variable (not sure if that's a thing).

**Comment by trstndvll** [comment]
There's a function in gitCommands.ts that runs all the git commands, called addCommand. This is what I use to call git cherry but it doesn't seem to produce any return.

**Comment by trstndvll** [comment]
![Screenshot 2019-05-24 21 55 14](https://user-images.githubusercontent.com/12837282/58364855-5c81f080-7e6f-11e9-9033-d87692e80a5c.jpg)
So I've been trying to use a child process to use the exec function, which lets you run a command and take in the output. I can get the 'git cherry' command working, but the problem is it isn't for the repo that we have open in the app, it's for the visualgit repo itself.  Now I'm thinking, we're going to have to take in the file path, execute a cd command so we can get to the local file path we want to check, and then run 'git cherry.'

Does anyone have any ideas? I'm not sure if this is the best route to go. Is there another, more obvious way to check for commits that haven't been pushed?

**Comment by jerushachua** [comment]
'Git status' can be used to check whether or not your branch is up to date. If you have commits which haven't been pushed, it will say 'x number of commits ahead of HEAD' or something similar. 

If the VisualGit repo tracks the remote repo instead of the local repo, this still may not work.

**Comment by trstndvll** [comment]
Thanks for the suggestion @jerushachua ! Git status would totally work also. The output is a little less friendly than git cherry, just in terms of reading it afterwards from a variable (cherry returns just the IDs of the commits). But you're right, either way we have to see if there's a way to execute the git command inside the local repo.

**Comment by nikola-rados** [comment]
@trstndvll It looks like in `git.ts` the `refreshAll(...)` function actually looks through the opened repo for commits:
```
...
          Git.Reference.nameToId(repository, branchList[i].name()).then(function (oid) {
            // Use oid
            console.log("old id " + oid);
            if (branchList[i].isRemote()) {
              // for remote branches add oid and branch name to remote branches map
              remoteName[bp] = oid;
            } else {
              //add branch name to the branch list
              branchCommit.push(branchList[i]);
              console.log(bp + " adding to end of " + oid.tostrS());
              if (oid.tostrS() in bname) {
                bname[oid.tostrS()].push(branchList[i]);
              } else {
                bname[oid.tostrS()] = [branchList[i]];
              }
            }
...
```
The `console.log(bp + " adding to end of " + oid.tostrS());` line will show up in the logs with the id of the commit(s) that have yet to be pushed.  So we can already find the commits (I think)!
 
I've tried adding a function here that will set the `PushButNoCommit` flag if we find unpushed commits, however I ran into some issues with the implementation.  I think that we can make something like this work without having to add a whole new git command (whether it be cherry or status) since it looks as if the work is being done for us.

That being said I think we should explore further to get a better understanding of what is going on and if we can use it.

**Comment by trstndvll** [comment]
@nikola-rados That's awesome! Yes let's definitely look into this

**Comment by trstndvll** [comment]
@nikola-rados So if I understand the code correctly, I think it's just grabbing the last commit ID no matter if it's already been pushed. What I mean by that is if I have no commits and I try to push (with my check turned off), that oid variable is returning the last commit ID from my last push. So I'm not sure if that will work unfortunately since that variable won't ever be null (i.e. won't ever have an ID). If you understand it differently though let me know!

**Comment by nikola-rados** [comment]
Ah good point.  When I was testing it I happened to have a commit that I didn't push and expected to see.

**Comment by trstndvll** [comment]
@jtiz003 Hi James, my group and I had a question for you. We believe we've implemented a fix for the app crashing when it tries to push an empty commit, but we've found an edge case. We have a check in place that checks if there have been commits in the session, and will not allow you to push if there haven't been any commits. 

Here's the edge case: currently, if you commit changes in visualgit, then exit the app and re-open, the commits will not show and you cannot push because of our check. We believe we have solved the issue with the app crashing at least, but this edge case has presented quite a challenge as we have been trying to find a way to check for un-pushed commits using nodegit to check the repo. 

Our question to you is would you like the edge case fixed in this issue, or do you believe it's worth it to make another issue for sprint 2? Thank you for your input!

**Comment by jtiz003** [comment]
@trstndvll sorry for the slow reply, have you made any progress with the "edge case"?

It sounds like knowing the difference between the local and remote commit history/status would resolve this issue? Perhaps we can have a new feature that checks this and prompts the user to sync the local and remote

**Comment by Unimax** [comment]
I found out that the application is crashing because of the 'cred' object we are using for authentication in a callback if failing due to one of the following reason:

1. it's not able to create the credential object in the C layer. (at least for the 2nd time) .becuase with some changes i was able to see the exact error as in the issue https://github.com/nodegit/nodegit/issues/982
2. because we are reusing the object it's expired and hence having an issue. not sure why.

to prove that is the issue we can simply (I tested on Windows 10 and Ubuntu 18.04)
change the line 406 in master, or 416 in the issue30 branch i.e. 
`return cred;`
into 
`return Git.Cred.userpassPlaintextNew("username", "password");
`
Note: for now use ur username and password.

Possible Solutions:
1. I notice that we are using **nodegit** instead of **Git** in some places to generate the cred object. so we can change it to Git. more details on that https://github.com/nodegit/nodegit/issues/982
2. it will be fixed with the Oauth changes.

Current Flag 'pushButNoCommit' solves an application level optional choice that you want to let the user try to push without any local commits. I would say yes, especially when we don't have any way to detecting any local commit done before the start of the application. (by command line or by the same application). Also because with the Auth fixed it won't crash the application even if you push with no local commits. (tested!!)

@everyone can anyone else test this?

@colinwerner link this with OAuth issue.

also in linux, i have another issue with the branch name coming as "Branch: master" which I fixed by changing the line 391 (master) . (on Windows this was not a problem)
`let branch = document.getElementById("branch-name").innerText.split(": ")[1];
`
